src/test/scala/DSATest.scala

```diff
@@ -0,0 +1,707 @@
+
+  }
+
+  // this function asserts universal properties about global objects in Jumptable2  example
+  def assertJumptable2Globals(dsg: DSG) : Unit = {
+    // global mappings
+
+    // __libc_csu_init relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69600, 69608))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(2136, 2136 + 124))._1.cells(0))))
+    // __lib_csu_fini relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69560, 69568))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(2264, 2268))._1.cells(0))))
+    // jumptable relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69624, 69632))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(69656, 69656 + 24))._1.cells(0))))
+    // add_two relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69656, 69656 + 24))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(1948, 1948 + 36))._1.cells(0))))
+    // add_six relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69656, 69656 + 24))._1.cells(8).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(1984, 1984 + 36))._1.cells(0))))
+    // sub_seven relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69656, 69656 + 24))._1.cells(16).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(2020, 2020 + 36))._1.cells(0))))
+    // main relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69608, 69616))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(2056, 2056 + 76))._1.cells(0))))
+    // x relocation
+    assert(dsg.adjust(dsg.globalMapping(AddressRange(69584, 69584 + 8))._1.cells(0).getPointee).equals(dsg.find(dsg.globalMapping(AddressRange(69648, 69648 + 4))._1.cells(0))))
+  }
+
+  test("local jumptable2 callees") {
+    val results = RunUtils.loadAndTranslate(
+      BASILConfig(
+        loading = ILLoadingConfig(
+          inputFile = "examples/jumptable2/jumptable2.adt",
+          relfFile = "examples/jumptable2/jumptable2.relf",
+          specFile = None,
+          dumpIL = None,
+        ),
+        staticAnalysis = Some(StaticAnalysisConfig()),
+        boogieTranslation = BoogieGeneratorConfig(),
+        outputPrefix = "boogie_out",
+      )
+    )
+
+
+    val program = results.ir.program
+    // test that all three calles have the same local graph
+    val callees = Set("sub_seven", "add_two", "add_six")
+    callees.foreach(
+      callee =>
+        val dsg = results.analysis.get.locals.get(program.procs(callee))
+        assert(dsg.stackMapping.isEmpty) // stack is not used in either callee
+        assertJumptable2Globals(dsg) // globals should be the same everywhere unused in callees
+        // x should point to a collapsed object, in all 3 functions
+        // all three load value of x
+        // the analysis doesn't know if x is a pointer or not therefore assumes it is for soundness
```

Comment:
Can the analysis not tell that x is not a pointer because it is a 32 bit value and therefore not a pointer?

Reply:
In general:
- 64-bit values: possibly pointers
- 128-bit values: potentially two 64-bit pointers concatenated together
- anything else: definitely not a pointer